api/include/opentelemetry/trace/trace_state.h

```diff
@@ -0,0 +1,98 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <cstring>
+
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/nostd/unique_ptr.h"
+
+namespace opentelemetry
+{
+namespace trace
+{
+
+/**
+ * TraceState carries tracing-system specific context in a list of key-value pairs. TraceState
+ * allows different vendors to propagate additional information and inter-operate with their legacy
+ * Id formats.
+ *
+ * Implementation is optimized for a small list of key-value pairs.
+ *
+ * Key is opaque string up to 256 characters printable. It MUST begin with a lowercase letter,
+ * and can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+ * forward slashes /.
+ *
+ * Value is opaque string up to 256 characters, of printable ASCII RFC0020 characters (i.e. the
+ * range 0x20 to 0x7E) except comma , and equals =.
+ */
+class TraceState
+{
+public:
+  static constexpr int kKeyMaxSize       = 256;
+  static constexpr int kValueMaxSize     = 256;
+  static constexpr int kMaxKeyValuePairs = 32;
+
+  // Class to store key-value pairs.
+  class Entry
+  {
+  public:
+    Entry() noexcept = default;
+
+    // Creates an Entry for a given key-value pair.
+    Entry(nostd::string_view key, nostd::string_view value) noexcept;
+
+    const char *GetKey();
+    const char *GetValue();
+
+  private:
+    // Store key and value as raw char pointers to avoid using std::string.
+    const char *key_;
+    const char *value_;
+  };
+
+  // An empty TraceState.
+  TraceState() noexcept : entries_(new Entry[kMaxKeyValuePairs]), num_entries_(0) {}
+
+  // Returns false if no such key, otherwise returns true and populates value.
+  bool Get(nostd::string_view key, nostd::string_view value) const noexcept { return false; }
+
+  // Creates an Entry for the key-value pair and adds it to entries.
+  void Set(nostd::string_view key, nostd::string_view value) const noexcept;
+
+  // Returns true if there are no keys, false otherwise.
+  bool Empty() const noexcept { return true; }
+
+  // Returns a span of all the entries. The TraceState object must outlive the span.
+  nostd::span<Entry> Entries() const noexcept { return {}; }
+
+  // Returns whether key is a valid key. See https://www.w3.org/TR/trace-context/#key
+  static bool IsValidKey(nostd::string_view key);
+
+  // Returns whether value is a valid value. See https://www.w3.org/TR/trace-context/#value
+  static bool IsValidValue(nostd::string_view value);
+
+private:
+  // Store entries in a C-style array to avoid using std::array or std::vector.
+  nostd::unique_ptr<Entry[]> entries_;
```

Comment:
Can we still use a `unique_ptr` in that case? I wanted to use a `unique_ptr` to have the memory management of the array taken care of.

Reply:
If you have a plain array (`Entry entries_[32];`), you don't need to worry about memory management. The array memory will be freed when the owning `TraceState` object is freed, and the destructor for all `Entry`s will be called.

Furthermore, this will make it easier to define copy operations for `TraceState`.

Note that this is a `Entry entries_[32];`, and not a `Entry *entries_[32];`.

Reply:
I changed it to `Entry entries_[32]`, but I'm still curious about my comment above.